File: python/questions/maxProduct.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        ts = []
        def dfs(root: Optioal[TreeNode]) -> int:
            if root == None:
                return 0
            sum = root.val + dfs(root.left) + dfs(root.right)
            ts.append(sum)
            return sum
        total = dfs(root)
        ans = 0
        for ns in ts:
            ans = max(ans, ns*(total-ns))
        return ans % 1_000_000_007

File: cpp/questions/maxProduct.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxProduct(TreeNode* root) {
        std::vector<int> ts;
        auto dfs = [&] (this auto&& dfs, TreeNode* root) -> int {
            if (root == nullptr) {
                return 0;
            }
            int sum = root->val + dfs(root->left) + dfs(root->right);
            ts.push_back(sum);
            return sum;
        };
        long long total = dfs(root);
        long long ans = 0;
        for (int& cs : ts) {
            ans = std::max(ans, cs*(total-cs));
        }
        return ans%1'000'000'007;
    }
};
